### PR TITLE
[D1] Add generated columns docs

### PR DESCRIPTION
**Pull request by elithrar**
This PR:

- Adds a new `/d1/learning/generated-columns/` doc that shows how to define generated columns (https://www.sqlite.org/gencol.html), and why they are useful
- Cross-links to that from the Querying JSON docs, as generated columns + JSON functions are extremely powerful in D1
- Mentions the new docs in the changelog

### CODE REVIEW

**Comment by elithrar** [comment]
Also relates to https://github.com/cloudflare/workerd/pull/779

File: content/d1/learning/generated-columns.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+---
+title: Generated columns
+pcx_content_type: concept
+weight: 10
+---
+
+# Generated columns
+
+D1 allows you define generated columns: columns based on the values of one or more other columns, SQL functions, or even [extracted JSON values](/d1/learning/querying-json).
```
Comment (Maddy-Cloudflare) [inline content/d1/learning/generated-columns.md]:
```suggestion
D1 allows you to define generated columns based on the values of one or more other columns, SQL functions, or even [extracted JSON values](/d1/learning/querying-json).
```

File: content/d1/learning/generated-columns.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+Generated columns can also have [indexes defined](/d1/learning/using-indexes/) against them, which can dramatically increase query performance over frequently queried fields.
+
+## Types of generated columns
+
+There are two types of generated columns:
+
+* `VIRTUAL` (default): the column is generated when read. This has the benefit of not consuming storage, but can increase compute time (and thus reduce query performance), especially for larger queries.
+* `STORED`: the column is generated when the row is written. The column takes up storage space just as a regular column would, but the column does not need to be generated on every read, which can improve read query performance.
+
+When omitted from a generated column expression, generated columns default to the `VIRTUAL` type. The `STORED` type is recommended when the generated column is compute intensive: for example, parsing large JSON structures.
```
Comment (Maddy-Cloudflare) [inline content/d1/learning/generated-columns.md]:
```suggestion
When omitted from a generated column expression, generated columns default to the `VIRTUAL` type. The `STORED` type is recommended when the generated column is compute intensive. For example, when parsing large JSON structures.
```

File: content/d1/learning/generated-columns.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+## Defining a generated column
```
Comment (Maddy-Cloudflare) [inline content/d1/learning/generated-columns.md]:
```suggestion
## Define a generated column
```

File: content/d1/learning/generated-columns.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+-- Filter out "expired" results based on your generated column:
+-- SELECT * FROM your_table WHERE current_date() > expires_at
+ALTER TABLE your_table
+-- calculates a date (YYYY-MM-dd) 30 days from the current date.
+ADD COLUMN expires_at AS (date('now', '+30 days');
+```
+
+## Additional considerations
+
+* Tables must have at least one non-generated column: you cannot define a table with only generated column(s).
```
Comment (Maddy-Cloudflare) [inline content/d1/learning/generated-columns.md]:
```suggestion
* Tables must have at least one non-generated column. You cannot define a table with only generated column(s).
```

File: content/d1/learning/querying-json.md [hunk 5]
```diff
@@ -71,6 +72,24 @@ ERROR 9015: SQL engine error: query error: Error code 1: SQL error or missing da
...
+
+```sql
+CREATE TABLE some_table (
+    -- other columns omitted
+    -- JSON: {"measurement":{"aqi":[21,42,58],"wind_mph":"13","location":"US-NY"}}
+    location AS (json_extract(raw_data, '$.measurement.location')) STORED
+)
+```
+
+Visit the [generated columns documentation](/d1/learning/generated-columns/) to learn more about how to generate columns.
```
Comment (Maddy-Cloudflare) [inline content/d1/learning/querying-json.md]:
```suggestion
Refer to [Generated columns](/d1/learning/generated-columns/) to learn more about how to generate columns.
```

**Review by Maddy-Cloudflare — APPROVED** [review]

**Pull request merged**